### ISSUE TITLE
new data as a input

### ISSUE THREAD

**Issue by karaposu** [shivang8/Digit-Recognition#1]
Hello shivang8 , thats good coding. I am trying to figure out how to use new data as a input .

first problem is that i want to use 28x28 .bmp picture whichis created  with paint.  so i create function which  takes the image  converts data to logical array in format of  1x784 . here it is
```
`  function [s] = row_and_col_values_of_img(test_picture)
  s=zeros(1,784) ;
 I =(test_picture)  ;
          for j=1:28  
                 for i=1:28  
            s(1, i+ 28*(j-1) )=I(j,i); 
                 end
              end
 end ` 
```

since i used  ".bmp" format  with  one colour all my values are logical.   

then i used weights and biases to calculate output ..

```
nn = load('data_epoc200.mat');
test_folder = 'C:\Users\enesk\Desktop\w-s\nn-matlab\50piksel\28piksel'; % folder with big images
files_1 = dir(fullfile(test_folder));

Picture_set_1=imread(char(join( (test_folder)+"\"+files_1(3).name ) ) );
s = row_and_col_values_of_img(Picture_set_1);

weight1= nn.weight1;
weight2=nn.weight2;
bias1= nn.bias1;
bias2=nn.bias2;
neth = s * weight1;
neth = neth + bias1;
outh = sigm(neth);
% Output Layer
 neto = outh * weight2;
 neto = neto + bias2;
 outo = sigm(neto);
```

but the output values are not matching with the image label ....

my picture is 5.  

but the output array gives me  something like this 

0,00890543561907159	0,00137493916567926	0,000313196127740974	0,00203675665040161	2,50346713054253e-06	0,000663910983342224	0,000115502557563390	0,000172873298425822	0,956766243598225	0,000439554708199857

closest value to 1 is 9th values so it says my icture is 9 ....

**Comment by karaposu** [comment]
i just saw your prediction function  . can you explain these codes?
```
 % Error Phase
    min_error = 1:10;
    for i = 1:10
        v = targets(i,:);
        errora = v - outo;
        errorf = errora.^2;
        errorf = 0.5 * errorf;
        error = sum(errorf);
        min_error(i) = error;
    end
    [~,k] = min(min_error);
    k = k-1;
    %disp(min_error)
```

**Comment by karaposu** [comment]
and also prediction function is not producing right answers either....

**Comment by shivang8** [comment]
Can you please send me your .bmg image so that I can check for the output.

Regards

Shivang

**Comment by shivang8** [comment]
I will break down the code for simplified explination:

% Error Phase
    min_error = 1:10;

In the above step I have created an list from 1 to 10 this is done so
that we don't need to append the list again and again thus provide
better speed.


    for i = 1:10

Loop to iterate through all target array


        v = targets(i,:);

Our target array consist of binary target values
i.e. target = [ 1 0 0 0 0 0 0 0 0 0 ,        --> For Zero
                     0 1 0 0 0 0 0 0 0 0 ,        --> For One
                     0 0 1 0 0 0 0 0 0 0 ... ]    --> For Two and so on
We are using the ith wor of target matrix.

        errora = v - outo;

V is our target values for that particular i. and OUTO is the output
from the output layer.


        errorf = errora.^2;

Squaring the error elements indivisually as errorf is in matrix form.


        errorf = 0.5 * errorf;

Half of errorf.


        error = sum(errorf);

Finding the total error as sum of errorf


        min_error(i) = error;

Storing the error in the ith element of min_error array which we
initilized earlier.


    end

End of loop


    [~,k] = min(min_error);

Finding the index of element with minimum value.


    k = k-1;

Since indexing in matlab is done by 1 onwards so if the prediction is
1 then the digit is 0, if the prediction is 2 the digit is 1... ans so
on.


Hope this would help you.
If you are still not clear why it is done kindly proceed to this link
(https://mattmazur.com/2015/03/17/a-step-by-step-backpropagation-example/)
this is the link I studied back propogation from.


Regards


Shivang Srivastava

**Comment by shivang8** [comment]
Sir the prediction function is giving 97% accuracy after running 200
iterations through the dataset i.e. 200 epocs. Here is the link to the
dataset of the weights after running 200 iterations
https://github.com/shivang8/Digit-Recognition/blob/master/data_epoc200.mat
kindly run your outputs on this for best result.

Kindly send the case where this fails so that I am able to improvise this.

Thanking You

Shivang

**Comment by karaposu** [comment]
Thank you for your response ;  i am using your preprocessed network also ..  the only thing i am doing different is using external images.  (still mnist data tho)

**Comment by karaposu** [comment]
```
clear all  
close all
clc

% %%   Begin training -  NO NEED SINCE THERE IS PREPROCESSED DATA ..
% 
% % Calling Initilization.m Script
% initilization
% 
% % Calling Training.m Script
% for epoc = 1:200
%     
%     msg=['new epoch number : ',num2str(epoc)];
%     msg2=['training continues...'];
%     disp(msg)
%        disp(msg2)
%     alpha = 1/(epoc^0.5);
%     %alpha = 0.5;
%     training
% 	
% 	
% end
% 
% Accuracy
% 
% save 'C:\Users\acer\Desktop\Digit Recognition\data_epoc200.mat'
%%   Ready data load

nn = load('data_epoc200.mat');
%neural_net = nn.net;

% 
% Picture_set_1=imread(char(join( (test_folder)+"\"+files_1(4).name ) ) );
% Picture_set_2=imread(char(join( (test_folder)+"\"+files_1(6).name ) ) );
% 
% s = row_and_col_values_of_img(Picture_set_1);
% 
% s2 = row_and_col_values_of_img(Picture_set_2);

% t1=binary_convertor(s2) ;



%% lOAD External TEST PIC HERE 


test_folder = 'C:\Users\enesk\Desktop\w-s\nn-matlab\50piksel\28piksel'; % test image folder on my computer... you need to change this to where your image is..
files_1 = dir(fullfile(test_folder));   % list files in that directory...

Picture_set_1=imread(char(join( (test_folder)+"\"+files_1(4).name ) ) );    % open files_1 in matlab workspace and find out your image's order
                                                                            %in directory.because directory  has items like "." and ".." so it starts from 3.  Check it out
imshow(Picture_set_1)  %show the image  
  
t = binary_convertor(Picture_set_1)   %use it if it is a non binary image . i have tried with non binary and binary images. same result. 

s = row_and_col_values_of_img(t);  %this is my function to create  1x784 vector of image piksels. this is how mnist data was created. it says row-wise in this official site...


% get what you need to use prediction function .
weight1= nn.weight1;
weight2=nn.weight2;
target=nn.target;
bias1= nn.bias1;
bias2=nn.bias2;

 %% prediction
    
     %s=~s ;
    k = prediction(s, weight1, weight2, bias1, bias2, target)
```

**Comment by karaposu** [comment]
```
%This is my function  for calculating 1x724  vector of image  

function [s] = row_and_col_values_of_img(test_picture)

   s=zeros(1,784) ;
    I =(test_picture)  ;
          for j=1:28  %
                for i=1:28  %
             s(1, i+ 28*(j-1) )=I(j,i); 
                 end

          end
            
          
end

% 
% quick check to see if it is truly calculates ...
%     
% at first  j =1   
%     
%                 for i=1:28  %
%              s(1, i )=I(j,i); 
%                  end
% so it just copy first row to s   ...
%     
% then   j =2
%                  for i=1:28  %
%              s(1, i+ 28) )=I(2,i); 
%                  end
%                  
%     so it copies second row to  s ; starting from s(1, 29)  until  s(1, 56)         
%     
%     
%     then   j =28
%     
%      for i=1:28  %
%              s(1, i+ 28*(27) )=I(28,i); 
%                  end
%                  
%   it copies 28th row to s  starting from s(1, (28*27) -1)  until  s(1, 28*28)...
    
```

**Comment by karaposu** [comment]
now you can import external picture and  predict  it is value..  here some of  my images :
![p3](https://user-images.githubusercontent.com/31979644/36015602-3944c5ea-0d80-11e8-9287-5e8194a63883.png)

![3](https://user-images.githubusercontent.com/31979644/36015605-42897b50-0d80-11e8-8e4f-fa2cb7d04835.png)

![p2](https://user-images.githubusercontent.com/31979644/36015606-42d460f2-0d80-11e8-95bc-e928738c03bc.png)

**Comment by karaposu** [comment]
cant upload .bmp files directly to github so i uploaded it  to this link

https://uploadfiles.io/bq1t4

**Comment by shivang8** [comment]
Thanks for pointing it out. I am working on it will respond at the earliest.

**Comment by shivang8** [comment]
With respect to the image you provided  i.e. https://uploadfiles.io/bq1t4
The thickness of the image is too thin which does not match our training data.  As the MNSIT dataset consists of atleast 45% of the character in the entire image. 

This is the improved version of your image https://ufile.io/nt3iu 

This may also happen due to the reason that the Neural Network is trained by a particular dataset thus it does not have enough knowledge of multiple handwritings. You need to train it with more dataset for best results.

If you have any additional dataset kindly send it to me for effective training.

**Comment by karaposu** [comment]
i have data set but merging it mnist is not very usefull i think . Since they have different charasterictics  they cancel each other weight's out.   I will upload them to my repo. U can try them if you like

https://github.com/karaposu/Digit_Recognition_neural_network_4digitonly



________________________________
Gönderen: Shivang Srivastava <notifications@github.com>
Gönderildi: 12 Şubat 2018 Pazartesi 13:51
Kime: shivang8/Digit-Recognition
Bilgi: Enes; Author
Konu: Re: [shivang8/Digit-Recognition] new data as a input (#1)


With respect to the image you provided i.e. https://uploadfiles.io/bq1t4<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuploadfiles.io%2Fbq1t4&data=02%7C01%7C%7Cffcb08b273b54eaabd8208d572069691%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636540294973977968&sdata=pwCw9XD8lepRb%2BTcJfVYUWwaR9XOG6LLeqtaFM47J9o%3D&reserved=0>
The thickness of the image is too thin which does not match our training data. As the MNSIT dataset consists of atleast 45% of the character in the entire image.

This is the improved version of your image https://ufile.io/nt3iu<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fufile.io%2Fnt3iu&data=02%7C01%7C%7Cffcb08b273b54eaabd8208d572069691%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636540294973977968&sdata=a%2FA4hG174500e%2B30FkcDet1cJqHcPAKqGptjiMD%2FxqQ%3D&reserved=0>

This may also happen due to the reason that the Neural Network is trained by a particular dataset thus it does not have enough knowledge of multiple handwritings. You need to train it with more dataset for best results.

If you have any additional dataset kindly send it to me for effective training.

**Comment by shivang8** [comment]
I would sure have look and thanks for pointing it out.
